src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers
+{
+    /// <summary>
+    /// Wraps a PipeWriter so you can start appending more data to the pipe prior to the previous flush completing.
+    /// </summary>
+    internal class ConcurrentPipeWriter : PipeWriter
+    {
+        // The following constants were copied from https://github.com/dotnet/corefx/blob/de3902bb56f1254ec1af4bf7d092fc2c048734cc/src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs
+        // and the associated StreamPipeWriterOptions defaults.
+        private const int InitialSegmentPoolSize = 4; // 16K
+        private const int MaxSegmentPoolSize = 256; // 1MB
+        private const int MinimumBufferSize = 4096; // 4K
+
+        private readonly object _sync = new object();
+        private readonly PipeWriter _innerPipeWriter;
+        private readonly MemoryPool<byte> _pool;
+        private readonly BufferSegmentStack _bufferSegmentPool = new BufferSegmentStack(InitialSegmentPoolSize);
+
+        private BufferSegment _head;
+        private BufferSegment _tail;
+        private Memory<byte> _tailMemory;
+        private int _tailBytesBuffered;
+        private int _bytesBuffered;
```

Comment:
This should be a long, though in practice I dunno if that will matter, you'd probably OOM first 😄

Reply:
Technically, you're right. We should probably change it to a long in StreamPipeWriter too.